File: 207. Course Schedule.cpp
class Solution {
public:
    // DFS: build graph and find circle, O(N2)time, O(N)space, 58.97%, ok
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<unordered_set<int>> adj(numCourses, unordered_set<int>());
        for(auto p: prerequisites)
            adj[p.second].insert(p.first);
        vector<int> visited(numCourses, 0);
        vector<int> currCheck(numCourses, 0);
        for(int node = 0; node < numCourses; node++){
            if(!visited[node] && findCycle(node, visited, currCheck, adj))
                return false;
        }
        return true;
    }
    
    bool findCycle(int node, vector<int>& visited, vector<int>& currCheck, vector<unordered_set<int>>& adj){
        if(visited[node]) return false;
        visited[node] = currCheck[node] = 1;
        for(auto neighbor: adj[node]){
            if(currCheck[neighbor] || findCycle(neighbor, visited, currCheck, adj))
                return true;
        }
        currCheck[node] = 0;
        return false;
    }
};

File: 207. Course Schedule.py
class Solution(object):
    
    # topological sort: O(N) time, O(N) space
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        def findCycle(i, adjacent):
            if self.globalVisited[i]:
                return False
            self.globalVisited[i] = self.localVisited[i] = True
            for neigh in adjacent[i]:
                if self.localVisited[neigh] or findCycle(neigh, adjacent):
                    return True
            self.localVisited[i] = False  # back track 
            return False 
        
        if numCourses == 0 or len(prerequisites) == 0:
            return True
        self.globalVisited = [False for i in range(numCourses)]
        self.localVisited = [False for i in range(numCourses)]
        adjacent = {i:[] for i in range(numCourses)}
        for prev, curr in prerequisites:
            adjacent[prev] += [curr]
        for i in xrange(numCourses):
            if not self.globalVisited[i] and findCycle(i, adjacent):
                return False         
        return True
        